### ISSUE TITLE
Question: why doesnt state get reset after each epoch?

### ISSUE THREAD

**Issue by hughperkins** [karpathy/char-rnn#78]
Hi Andrej,

I'm still gradually struggling to figure out how LSTM works, so please excuse the naivety of my question.  Looking at the state, it looks like:
- it is zero'd at the start of training, https://github.com/karpathy/char-rnn/blob/master/train.lua#L152
- then gets updated after each `seq_length` sequence https://github.com/karpathy/char-rnn/blob/master/train.lua#L276
- ... and, I think that each character in each mini-batch, eg the first character in each mini-batch, walks over a contiguous set of `input_length / batch_size` characters from the input?
- and so, it seems to make sense to me that whilst the training is walking over these `input_length / batch_size` characters, that the state is persistent, as it is
- but it seems like, at the end of these characters, which corresponds to the end of each epoch, I think?, either the state should be zerod, or else each of the `batch_size` states should be rotated by one place, ie init_state[1][n] becomes init_state[1][b+1], for each b in `batch_size`?  What is missing from my understanding of how this works?

**Comment by karpathy** [comment]
Hey, nothing is missing you're right. The init state should be init state when we wrap to new epoch. This bug is negligeable in most cases and I don't think it introduces too much loss or confusion to most models. We could fix it with a pull request and probably expect ~0.01 better results on smaller datasets that contain few batches.

**Comment by hughperkins** [comment]
Ok cool.  Yes, I'm not claiming this will give any increased accuracy, just trying to check my understanding of how it works.

Slightly related, but similar idea, if I train a couple of epochs of 'The quick brown fox jumps over the lazy dog', with seq_length=3, batch_size=4, here is what get trained at each iteration:
```
x=The| br|x j|ver|	y=he |bro| ju|er |	
t=1 thisinput[1]	T xv	
t=2 thisinput[1]	hb e	
t=3 thisinput[1]	erjr	
x= qu|own|ump| th|	y=qui|wn |mps|the|	
t=1 thisinput[1]	 ou 	
t=2 thisinput[1]	qwmt	
t=3 thisinput[1]	unph	
x=The| br|x j|ver|	y=he |bro| ju|er |	
t=1 thisinput[1]	T xv	
t=2 thisinput[1]	hb e	
t=3 thisinput[1]	erjr	
x= qu|own|ump| th|	y=qui|wn |mps|the|	
t=1 thisinput[1]	 ou 	
t=2 thisinput[1]	qwmt	
t=3 thisinput[1]	unph	
```
- each sectoin starting `x=` is one iteration, where `x=` is the 4 sequences, separated by `|`, and
- each line `t=1`, `t=2` is one mini-batch, with the data sent to that mini-batch, ie 4 letters each time

Let's call the index into the mini-batch `b`, so for each mini-batch, b will take values 1,2,3,4.  Now, if we slice vertically downwards through one single value of b, eg b=1, we get the training using one set of state.  It looks like for b=1, for the first two epochs:

```
The quThe qu
```

It means that:
- the state associated with each value of b is only exposed to the same bit of training data, repeatedly
- Let's call the data associated with one b-value a `chunk`.  Any sequences in the data that cross between the breaks between chunks cannot be learned I think?

Perhaps it could be reasonable to do one of the following?
- rotate the states at the end of each epoch, so that b=2 receives the state from b=1, b=3 receives state from b=2, and so on, or
- rotate the training data, so that b=1 gets at epoch 2 the next data in the sequence, ie in this case 'ick br', and similar for b=2 will receive 'fox ju', and so on

What do you think?

**Comment by karpathy** [comment]
Yep, once the last batch is processed the current_state contents should be "cycled" downwards, and the first one (which should hold the beginning of the dataset) should be set to the initial state. I think.

If you're willing to make sure this interpretation is correct and add the cycle code I'd be happy to merge it in.

**Comment by hughperkins** [comment]
I dont think I'm going to implement this.  I was just trying to clarify my understanding really :-)

=> Closed.
